### ISSUE TITLE
nixos-generate-config doesn't "respect" mount options

### ISSUE THREAD

**Issue by Lillecarl** [NixOS/nixpkgs#146177]
#### Backstory
I got a System76 Lemur Pro from work which I installed NixOS on as soon as i got it and it's been running great, but every now and then I would have issues with coreboot not recognizing my boot drive(s). I mirror boot between 2 nvme disks.

I opened a System76 support ticket, but they have a hard time supporting a system that isn't Ubuntu or Pop!OS, which is reasonable.

So i was thinking, why have i never had this issue with my desktop that's also running NixOS, what's the issue?

I figured it might be hibernation, but I couldn't be certain. And since i run both my kernels and EFI on fat32 (oops, my fault) i figured it might be related to this particularly simple filesystem not behaving as I want it to. And since we NixOS people rebuild Grub a lot more than most other people i figured i just wrap nixos-rebuild switch into a bash function to call sync on /boot and /boot-fallback.

Now the issues disappeared, and I was happy for awhile. But then I figured, this can't be how this should be solved, I mean really?

So i looked into this a bit more and found a vfat mount option called "flush" which supposedly flushes data to disk sooner than normal. Which i guess could be important for a filesystem like fat32. Then i removed my wrapper function and kept using NixOS, since then i haven't lost my bootloader a single time (Previously when I lost it i booted a NixOS USB, mounted everything back up manually and did a nixos-install, rebooted and it all worked again.

Now i have no hard proof that this was the problem, but I haven't experienced the issue since and that's good enough for me.

I also tried setting the installasremoveable boot setting, which supposedly should make coreboot ignore efi variables and boot straight away as soon as it finds the efi file, this didn't help either which further supports my hypothesis.

So today i figured, since i added the flush mount option to my nixos config I'd see what ```nixos-generate-config --dir=/tmp``` does, and it seems like it doesn't respect my "flush" option.

#### The bug(maybe)
nixos-generate-config doesn't respect mount options, not sure if this is a feature or not. But for inexperienced users like myself this was a PITA until i figured "all of this" out.

#### Considerations
What makes me question myself is why I can't find any issues related to this, I imagine loads of people run NixOS on their laptop with a setup somewhat similar to mine (because they didn't know better when partitioning they accidentally but their initrd on fat32 along with the efi). But I can't seem to find any issues related to bootloaders disappearing out of nowhere.

I do have 40GB RAM (weird setup, 8GB soldered, 32GB DIMM). So hibernation sometimes takes awhile and keeps the disks quite busy. But I imagine loads of people are more extreme than myself here. (Funny side note, the system usually boots faster cold than resume from hibernation)

### Steps To Reproduce
1. Mount filesystems with some options
2. nixos-generate-config --dir=/tmp
3. cat /tmp/hardware-configuration.nix | grep "fileSystems\." -A 4
4. Doesn't have my mount options

### Expected behavior
Well, I would expect it to respect my mount options, after looking at /proc/self/mountinfo i realize that this might not be very simple, it's quite the string for each filesystem entry.

I would go as far as suggesting that the efi partition (whichever it is) is always generated with the flush option to help beginners (like myself) not go looking for their bootloaders.

### Additional context
https://gist.github.com/Lillecarl/77fff953aa6e4fb0f4c66413eb3eb7ab configuration.nix

https://gist.github.com/Lillecarl/677b2fc0def139e9c40ecf3395e21a5d hardware-configuration.nix

### Notify maintainers

@worldofpeace 
@Mic92 
@matthewbauer 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.78, NixOS, 21.05.4114.195d5816cdd (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"unstable-21.11pre330462.2fa862644fc, nixos-hardware, nixos-21.05.4114.195d5816cdd"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  nixos-install-tools
# a list of nixos modules affected by the problem
module:
```

**Comment by Mic92** [comment]
Did this happen with grub? In systemd-boot-builder I added `fsyncfs` to flush as fast as possible after new boot entries and kernels are written: https://github.com/NixOS/nixpkgs/blob/6b68d74f036def3d48a23969edfd0fa47f34af3e/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L300

**Comment by Mic92** [comment]
Scraping mount options from `/proc/mounts` is quite hard. I don't think we can just take all mount options from there, because not all will be valid.

**Comment by Mic92** [comment]
Also I would recommend for grub to use a boot filesystem different from EFI i.e. ext4 instead of fat32, which should be much safer to operated. So that EFI only contains grub and the actual `/boot` would store kernels and boot entries on a sane filesystem.

**Comment by Lillecarl** [comment]
Hi @Mic92!

Yes this has indeed been happening with grub, I couldn't find options to mirror systemd-boot to multiple drives as you can with grub so i "had" to choose grub.

I should also indeed change my partition layout so that the kernels don't reside on the efi partition. With that said I also think since NixOS is pretty much "parition as you like" when installing (No partition walkthrough) many people will have this because they didn't quite get the memo regarding this when they set their systems up (like myself).

I lost my bootloader at least once a day until i deployed the script, then the mount option. And now I've been going weeks w/o losing it, so I'm as certain as I can be (which isn't that much i guess, me being a relative beginner and all).

I will swap the grub partition for ext4 one drive at a time, but I would consider this still something that's broken.

**Comment by Mic92** [comment]
Yeah, we should have ext4 as a best practice for `/boot` even with EFI being used in our manual but still make sure we flush the filesystem in our grub boot generator script. There are still some use cases where grub is inevitable (BIOS-only server, encrypted or mirrored /boot, some broken arm firmware).

**Comment by Lillecarl** [comment]
@Mic92 I don't speak(or read) Perl but searching for flush or sync in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/loader/grub/install-grub.pl yields no results.

@edolstra
@nh2 

Seems like you guys have touched quite a bit on this script throughout the years, would either your you mind having a look at flushing both the efi and boot partitions when installing grub like the systemd-boot install script does? :)

**Comment by Artturin** [comment]
not a bug https://github.com/nixos/nixpkgs/commit/a146fdab808ddacc2ea0b950d753807d75eae2db

**Comment by iamevn** [comment]
then should `hardware-configuration.nix` be kept unedited from `nixos-generate-config` and have desired mount options overridden in `configuration.nix`?
